folding-schemes/src/folding/nova/decider_circuits.rs

```diff
@@ -0,0 +1,577 @@
+use ark_r1cs_std::{
+    alloc::AllocVar,
+    boolean::Boolean,
+    eq::EqGadget,
+    fields::{fp::FpVar, FieldVar},
+    groups::GroupOpsBounds,
+    prelude::CurveVar,
+    ToConstraintFieldGadget,
+};
+use ark_relations::r1cs::{ConstraintSynthesizer, ConstraintSystemRef, SynthesisError};
+use ark_std::Zero;
+use core::marker::PhantomData;
+
+use super::{
+    circuits::{ChallengeGadget, CommittedInstanceVar},
+    decider_eth_circuit::{KZGChallengesGadget, R1CSVar, RelaxedR1CSGadget, WitnessVar},
+    nifs::NIFS,
+    CommittedInstance, Nova, Witness,
+};
+use crate::arith::r1cs::R1CS;
+use crate::commitment::CommitmentScheme;
+use crate::folding::circuits::{
+    cyclefold::{
+        CycleFoldCommittedInstance, CycleFoldCommittedInstanceVar, CycleFoldConfig,
+        CycleFoldWitness,
+    },
+    nonnative::{affine::NonNativeAffineVar, uint::NonNativeUintVar},
+    CF1, CF2,
+};
+use crate::folding::nova::NovaCycleFoldConfig;
+use crate::frontend::FCircuit;
+use crate::utils::vec::poly_from_vec;
+use crate::Error;
+
+/// Circuit that implements part of the in-circuit checks needed for the offchain verification over
+/// the Curve2's BaseField (=Curve1's ScalarField).
+#[derive(Clone, Debug)]
+pub struct DeciderCircuit1<C1, C2, GC2>
+where
+    C1: CurveGroup,
+    C2: CurveGroup,
+    GC2: CurveVar<C2, CF2<C2>>,
+{
+    _c1: PhantomData<C1>,
+    _c2: PhantomData<C2>,
+    _gc2: PhantomData<GC2>,
+
+    /// E vector's length of the Nova instance witness
+    pub E_len: usize,
+    /// E vector's length of the CycleFold instance witness
+    pub cf_E_len: usize,
+    /// R1CS of the Augmented Function circuit
+    pub r1cs: R1CS<C1::ScalarField>,
+    pub poseidon_config: PoseidonConfig<CF1<C1>>,
+    /// public params hash
+    pub pp_hash: Option<C1::ScalarField>,
+    pub i: Option<CF1<C1>>,
+    /// initial state
+    pub z_0: Option<Vec<C1::ScalarField>>,
+    /// current i-th state
+    pub z_i: Option<Vec<C1::ScalarField>>,
+    /// Nova instances
+    pub u_i: Option<CommittedInstance<C1>>,
+    pub w_i: Option<Witness<C1>>,
+    pub U_i: Option<CommittedInstance<C1>>,
+    pub W_i: Option<Witness<C1>>,
+    pub U_i1: Option<CommittedInstance<C1>>,
+    pub W_i1: Option<Witness<C1>>,
+    pub cmT: Option<C1>,
+    pub r: Option<C1::ScalarField>,
+    /// CycleFold running instance
+    pub cf_U_i: Option<CycleFoldCommittedInstance<C2>>,
+
+    /// KZG challenges
+    pub kzg_c_W: Option<C1::ScalarField>,
+    pub kzg_c_E: Option<C1::ScalarField>,
+    pub eval_W: Option<C1::ScalarField>,
+    pub eval_E: Option<C1::ScalarField>,
+}
+impl<C1, C2, GC2> DeciderCircuit1<C1, C2, GC2>
+where
+    C1: CurveGroup,
+    <C1 as CurveGroup>::BaseField: PrimeField,
+    <C1 as Group>::ScalarField: Absorb,
+    C2: CurveGroup,
+    GC2: CurveVar<C2, CF2<C2>> + ToConstraintFieldGadget<CF2<C2>>,
+    for<'b> &'b GC2: GroupOpsBounds<'b, C2, GC2>,
+{
+    pub fn from_nova<GC1, CS1, CS2, const H: bool, FC: FCircuit<C1::ScalarField>>(
```

Comment:
Have you considered implementing the `TryFrom` trait instead?

Reply:
This is a good point. I'm checking the trait and seems to be for converting types:
*"This is useful when you are doing a type conversion that may trivially succeed but may also need special handling."*
(from https://doc.rust-lang.org/std/convert/trait.TryFrom.html )

whereas here in this method we're not 'converting' a type into another one but 'building' the Circuit1/Circuit2 from the given Nova instance.
Like from the given Nova instance we compute some new parameters and build the circuit struct which is then returned, so I'm not sure if it fits into the concept of TryInto (from a conceptual level).

But I'm not a Rust expert so I'll invoke @CPerezz hoping to get some light.
Maybe we should call the method something like `build_from_nova` or similar? (in order to make more explicit that this is building the circuit struct from the nova instance)

Reply:
Not sure I understand the comment.
As far I understand
a) it makes sense to change it to fit in the `TryFrom` trait
b) but on the same time you mention that with the `TryFrom` it will be harder to find when using the API, whereas having the custom `from_nova` method would be easier to find and understand.
I'm missing something?

Reply:
Ohh! Ok I see hehe, thanks.
Hmm then I'm unsure. Looks like the `TryFrom` is the more 'pure-rust-style' but that on the same time adds complexity of usage. And additionally there is the mentioned rationale about the `TryFrom` being a 'type conversion', while here we're not strictly converting between types, but building 2 circuits from the Nova struct (computing new values in the process (the crossterms of the NIFS, the commitment challenges, evaluations of polynomials, etc)).
With all that in mind, I'm slightly inclined towards keeping it as it is (without `TryFrom`), but I'm open to changing it if we find stronger reasonings.
 
How do you see it @dmpierre ?